File: 2025/days/Day03.cs
namespace _2025.days;

public static class Day03
{
    public static void Run()
    {
        string[] banks = File.ReadAllLines("inputs/day03.txt");
        long[] totals = [0, 0];
        foreach (string bank in banks)
        {
            totals[0] += long.Parse(GetLargestPossibleNumber(bank, 2));
            totals[1] += long.Parse(GetLargestPossibleNumber(bank, 12));
        }
        Console.WriteLine($"Part 1: {totals[0]}");
        Console.WriteLine($"Part 2: {totals[1]}");
    }

    private static string GetLargestPossibleNumber(string bank, int digits)
    {
        if (digits == 0) return "";
        char leadingDigit = bank[..(bank.Length - (digits - 1))].Max();
        int idx = bank.IndexOf(leadingDigit);
        return leadingDigit + GetLargestPossibleNumber(bank[(idx + 1)..], digits - 1);
    }
}

File: 2025/days/day03.py
from pathlib import Path

def get_largest_possible_number(bank: str, digits: int) -> str:
    if digits == 0: return ""
    leading_digit = max(bank[:len(bank) - (digits - 1)])
    idx = bank.index(leading_digit)
    return leading_digit + get_largest_possible_number(bank[idx+1:], digits - 1)

banks = Path("inputs/day03.txt").read_text().splitlines()
print(f"Part 1: {sum(int(get_largest_possible_number(bank, 2)) for bank in banks)}")
print(f"Part 2: {sum(int(get_largest_possible_number(bank, 12)) for bank in banks)}")